File: prototype/iclean/components/slider/iclean_slider.py
from bokeh.core.properties import Tuple, String, Int
from bokeh.layouts import column
from bokeh.models import Slider
from bokeh.io import show

class ICleanSlider(Slider):
    """
    Custom extension of Bokeh Slider that implements a websocket connection. For information on 
    Bokeh slider see,
        https://docs.bokeh.org/en/latest/docs/reference/models/widgets.sliders.html

    Args:
        Slider (Class): Basic numerical slider class from Bokeh.
    """
     

    __implementation__ = "iclean_slider.ts"

    socket = Tuple(String, Int, help="""
    A tuple describing the websocket connection info: (string: address, port: int)
    """)

File: casagui/bokeh/components/iclean_slider.ts
import { Slider, SliderView } from "models/widgets/slider"
import * as p from "core/properties"


/**
* Builds on Bokeh slider class to provide a websocket connected slider widget.
*
* @remarks
* This method is part of the {@link prototype/iclean/components/widgets/Slider | IClean}.
*
* @param socket - A tuple containing the websocket connection information: (address, port)
*
* @beta
*/

export class ICleanSliderView extends SliderView {
    model: ICleanSlider

    connect_signals(): void {
        super.connect_signals()
    }

    render(): void {
        console.log("render")
        console.log(this.model)
        super.render()
    }
}

export namespace ICleanSlider {
    export type Attrs = p.AttrsOf<Props>
    export type Props = Slider.Props & {
        socket: p.Property<[string, number]>,
    }
}

export interface ICleanSlider extends ICleanSlider.Attrs { }

export class ICleanSlider extends Slider {
    properties: ICleanSlider.Props

    constructor(attrs?: Partial<ICleanSlider.Attrs>) {
        super(attrs)
    }

    static init_ICleanSlider(): void {
        this.prototype.default_view = ICleanSliderView;

        this.define<ICleanSlider.Props>( ({String, Number, Tuple}) => ({
            socket: [ Tuple(String, Number) ],
        }))
    }
}